MRDL -> MRTK Dialog & Progress control

johnppella opened the pull request:
![mrtk_dialogprogress](https://user-images.githubusercontent.com/16143963/36229003-b4e43ff6-118a-11e8-9478-e7f794b07aa3.jpg)

# Overview
This pull request brings MRDL’s Dialog and Progress controls to MRTK. These are standard UI for the notification/information display and visual feedback for the data loading. **Both controls are tested and optimized for both HoloLens & Immersive headset.** Core scripts and prefabs are added to MRTK’s main folder under UX. Example scenes are added to MRTK’S Examples folder under UX.

# Supported devices

- HoloLens 
- Immersive headset 

# Demo video
- Dialog: https://gfycat.com/ConventionalDirtyKiskadee
- Progress: https://gfycat.com/JaggedDimLacewing

# Technical details 
You can find technical details in these README pages: 
- [DialogsExample README](https://github.com/johnppella/MixedRealityToolkit-Unity/blob/Feature_UX_Dialog_Progress/Assets/MixedRealityToolkit-Examples/UX/Readme/README_DialogsExample.md)
- [ProgressExample README
](https://github.com/johnppella/MixedRealityToolkit-Unity/blob/Feature_UX_Dialog_Progress/Assets/MixedRealityToolkit-Examples/UX/Readme/README_ProgressExample.md)- 

# Example Scenes
- [Dialog Example scene](https://github.com/johnppella/MixedRealityToolkit-Unity/blob/Feature_UX_Dialog_Progress/Assets/MixedRealityToolkit-Examples/UX/Scenes/DialogsExample.unity)
- [Progress Example scene](https://github.com/johnppella/MixedRealityToolkit-Unity/blob/Feature_UX_Dialog_Progress/Assets/MixedRealityToolkit-Examples/UX/Scenes/ProgressExamples.unity)

![mrtk_dialog_devices](https://user-images.githubusercontent.com/16143963/36229002-b4c86970-118a-11e8-88d8-23db55957fb2.jpg)

![mrtk_progress2](https://user-images.githubusercontent.com/16143963/36229001-b4b0ec00-118a-11e8-86c1-6e49cc208380.jpg)

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/MixedRealityToolkit-Unity?pullRequest=1718) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/MixedRealityToolkit-Unity?pullRequest=1718) before we can accept your contribution.<br/><br/>:x: johnppella [sign now](https://cla.opensource.microsoft.com/Microsoft/MixedRealityToolkit-Unity?pullRequest=1718)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/MixedRealityToolkit-Unity?pullRequest=1718) it.</sub>

cre8ivepark commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,161 @@
+        // Use this for initialization
+        void Start()
+        {
+            //if Immersive headset is used make buttons twice as big.
```
@keveleigh Do we have a standard way in MRTK to identify HoloLens vs Immersive headset? I remember SceneContentAdjuster uses XRDevice.GetTrackingSpaceType() == TrackingSpaceType.RoomScale or .Stationary but could we use it to differentiate the HoloLens & immersive headset?

StephenHodgson replied:
`HolographicSettings.isOpaque` should be enough to evaluate the platform.

cre8ivepark replied:
@StephenHodgson Thanks, Stephen! Yes, that was what I was looking for.

cre8ivepark replied:
Updated the script to use HolographicSettings.IsDisplayOpaque.

cre8ivepark commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicator.cs:
```diff
@@ -0,0 +1,315 @@
+                return;
+
+            //Check if immersive headset is used. if so, scale up the progress indicator by a factor of 2.
+            bool isHoloLensUsed = MixedRealityToolkit.Common.CameraCache.Main.clearFlags != CameraClearFlags.Skybox;
```
@keveleigh Same question here.

cre8ivepark replied:
Updated the script to use HolographicSettings.IsDisplayOpaque.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/Prototyping/Shaders/UnlitTransparentTriplanar.shader:
```diff
@@ -52,9 +54,10 @@ Shader "MixedRealityToolkit/Examples/UnlitTransparentTriplanar"
 			float1 _LightIntensity;
 			float1 _CameraIntensity;
-			UNITY_INSTANCING_CBUFFER_START(Props)
+			UNITY_INSTANCING_BUFFER_START(Props)
```
Is this compatible with previous versions of unity? 
Eg 5.6, 2017.1?

DDReaper replied:
I thought the plan was to only support from 2017.2 from now.  But as this is targeting Dev, should it matter?

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/SpatialMapping/Scripts/PlaneFindingTest.cs:
```diff
@@ -112,7 +112,7 @@ private void OnDrawGizmosSelected()
                     UnityEditor.Handles.DrawLine(corners[1], corners[3]);
                     UnityEditor.Handles.DrawLine(corners[2], corners[3]);
-                    UnityEditor.Handles.ArrowHandleCap(0, center, Quaternion.FromToRotation(Vector3.forward, normal), 0.4f, EventType.ignore);
+                    UnityEditor.Handles.ArrowHandleCap(0, center, Quaternion.FromToRotation(Vector3.forward, normal), 0.4f, EventType.Ignore);
```
This change should be reverted as well

cre8ivepark replied:
@StephenHodgson I remember that I had the issue on previous Unity version. @johnppella It seems like Unity made an automatic update on this. This gave me error messages on Unity 2017.2

DDReaper replied:
As dev is only going to support 2017.2+, I think this should be ok?

StephenHodgson replied:
Auto unity updates should handle these changes. Should leave editor version as is

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Resources/Dialog.prefab:
```diff
@@ -0,0 +1,2459 @@
+%YAML 1.1
```
Please move anything in resources folder to appropriate place. Resources folder was removed.

StephenHodgson commented on ProjectSettings/ProjectVersion.txt:
```diff
@@ -1 +1 @@
-m_EditorVersion: 2017.2.1f1
+m_EditorVersion: 2017.3.0f3
```
Please revert editor version.

DDReaper replied:
Agreed, Editor version should stay at 2017.2 until 2017.3.x is approved for use

johnppella commented:
Addressed: 
shader issue BUFFER->CBUFFER, Removed Resources folder, removed use of Resources.Load(), version changes

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/DialogExamples.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public class DialogExamples : MonoBehaviour
+    {
+        void Start()
```
nit: missing private modifier

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,155 @@
+        public string FinishMessage = "Finished!";
+
+        // Use this for initialization
+        void Start()
```
nit missing private modifier.  Also please remove default comment.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Profiles/DefaultButtonIconProfileTexture.asset:
```diff
@@ -11,7 +11,7 @@ MonoBehaviour:
   m_Script: {fileID: 11500000, guid: 235c7917337f78f4b853694125a79865, type: 3}
   m_Name: DefaultButtonIconProfileTexture
   m_EditorClassIdentifier: Assembly-CSharp:HUX.Buttons:ButtonIconProfileTexture
-  _IconNotFound_: {fileID: 2800000, guid: 2adc8135dd936004db37d5cead85f951, type: 3}
```
Do we want to remove this?

StephenHodgson replied:
@Railboy , @cre8ivepark is this a valid change?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/FadeObject.cs:
```diff
@@ -0,0 +1,126 @@
+using MixedRealityToolkit.InputModule;
+
+#if UNITY_WSA || UNITY_STANDALONE_WIN
+using UnityEngine.Windows.Speech;
```
Are we using this anywhere?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicator.cs:
```diff
@@ -0,0 +1,315 @@
+
+namespace MixedRealityToolkit.UX.Progress
+{
+    public abstract class Singleton<T> : MonoBehaviour where T : Singleton<T>
```
We're adding another type of Singleton into the toolkit?

Not sure this is a good idea.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicator.cs:
```diff
@@ -0,0 +1,315 @@
+using UnityEngine;
+using UnityEngine.XR.WSA.Input;
+using MixedRealityToolkit.Utilities.Solvers;
+using UnityEngine.XR.WSA;
```
Doesn't compile for standalone or other platforms.

johnppella commented:
Added private modifier to Start() in DialogExamples.cs and ProgressExamples.cs
removed default comment from Start() 
Remoived unused usings from FadeObject.cs
Removed local definition of Singleton in ProgressIndicator.cs - replaced with MixedRealityToolkit.Common.Singleton

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicator.cs:
```diff
@@ -0,0 +1,274 @@
+                    else
+                    {
+                        // The prefab sent in the function overrides the prefab set in the inspector 
+                        // instantiatedCustomObject = GameObject.Instantiate((prefab == null) ? defaultPrefab : prefab, animator.transform) as GameObject;
```
why is this commented out?

Should we remove it?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressButton.cs:
```diff
@@ -0,0 +1,23 @@
+
+        private void OnButtonClicked(GameObject obj)
+        {
+            string na = this.gameObject.name;
```
Why are we getting the name?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ObjectDisplayScript.cs:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    // Update is called once per frame
+    void Update()
```
nit private access modifier. We can remove the comment above as well.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ObjectDisplayScript.cs:
```diff
@@ -0,0 +1,65 @@
+public class ObjectDisplayScript : MonoBehaviour
+{
+    [Header("How fast does object rotate?")]
+    public float rotationIncrement;
```
We should probably encapsulate these fields with properties if they need to be accessed by an external class and make these private and serialize the fields.

StephenHodgson replied:
Also, use PascalCasing for public and camelCasing for private

StephenHodgson replied:
(Goes for all classes in this PR.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/LoadingAnimation.cs:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    // Update is called once per frame
+    void Update()
```
nit missing private access modifier and we can remove the comment above

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/LoadingAnimation.cs:
```diff
@@ -0,0 +1,220 @@
+    private LoadingAnimation loadingAnimation;
+
+    // Use this for initialization
+    void Start()
```
nit missing private access modifier and we can remove the comment above

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/FadeObject.cs:
```diff
@@ -0,0 +1,118 @@
+        }
+
+        // Update is called once per frame
+        void Update()
```
nit missing private access modifier and we can remove the comment above

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/FadeObject.cs:
```diff
@@ -0,0 +1,118 @@
+        private void Awake()
+        {
+            mRenderer = this.GetComponent<Renderer>();
+            mCachedColor = mRenderer.material.color;
```
Memory leak in this class when calling renderer.material (here and below).

Please properly cache the material and dispose of and created instances.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogShell.cs:
```diff
@@ -0,0 +1,215 @@
+                        pos += len;
+
+                        while (pos < endOfLine && Char.IsWhiteSpace(text[pos]))
+                            pos++;
```
strange brace formatting in this section.  (also missing braces on this line.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogShell.cs:
```diff
@@ -0,0 +1,215 @@
+                if (endOfLine == -1)
+                    next = endOfLine = text.Length;
+                else
+                    next = endOfLine + System.Environment.NewLine.Length;
```
missing braces

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogShell.cs:
```diff
@@ -0,0 +1,215 @@
+using System.Collections.Generic;
+using System.Text;
+using UnityEngine;
+using UnityEngine.XR.WSA;
```
breaks cross platform compatibility

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogButton.cs:
```diff
@@ -0,0 +1,36 @@
+
+        public Dialog.Dialog.ButtonTypeEnum ButtonTypeEnum;
+
+        void OnEnable()
```
missing private access modifier

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/Dialog.cs:
```diff
@@ -0,0 +1,255 @@
+        /// </summary>
+        protected abstract void SetTitleAndMessage();
+
+        //todo: is this necessary?
```
Is it?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/Dialog.cs:
```diff
@@ -0,0 +1,255 @@
+        protected void Launch(SimpleDialogResult newResult)
+        {
+            if (state != StateEnum.Uninitialized)
+                return;
```
missing braces

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/Dialog.cs:
```diff
@@ -0,0 +1,255 @@
+    /// Can be extended to include more information for dialog construction
+    /// (eg detailed messages, button names, colors etc)
+    /// </summary>
+    public class SimpleDialogResult
```
please move into it's own file and encapsulate the fields.

johnppella commented:
more Changes- added missing braces, 
FadeObject.cs cached material instead of renderer
added private modifier in several places

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/FadeObject.cs:
```diff
@@ -100,19 +100,28 @@ void Update()
+
+        public void OnDestroy()
+        {
+            DestroyImmediate(mCachedMaterial);
```
`DestroyImmediate` should only be used for editor code only.  Please use `Destroy` instead.

johnppella commented:
changed script files.
deleted commented code

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/Prototyping/Shaders/UnlitTransparentTriplanar.shader:
```diff
@@ -1,89 +1,89 @@
+ ///
+ /// Simple vertex shader that blends static lighting and camera lighting.
+ ///
+	Shader "MixedRealityToolkit/Examples/UnlitTransparentTriplanar"
```
Some strange space formatting changes in this file.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/DialogExamples.cs:
```diff
@@ -0,0 +1,21 @@
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+using UnityEngine.XR.WSA;
```
breaks cross platform compilation

cre8ivepark replied:
Updated to use #ifdef following other existing code examples.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/DialogExamples.cs:
```diff
@@ -0,0 +1,21 @@
+            // Optimize the content for immersive headset
+            if (HolographicSettings.IsDisplayOpaque)
+            {
+               GameObject buttonCollection = GameObject.Find("ButtonCollection");
```
Where is this button in relation to this GameObject the component will be attached to?

cre8ivepark replied:
The buttons are under ButtonCollection object in the scene hierarchy.
![2018-02-26 17_17_29-unity 2017 2 1f1 personal 64bit - dialogsexample unity - mixedrealitytoolkit-u](https://user-images.githubusercontent.com/13754172/36705290-3063199e-1b19-11e8-9d2c-329c59a9d876.png)

StephenHodgson replied:
Can we get rid of the find method here and just reference it in the inspector? 

The Find should only be a fallback if the inspector value is null.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/LaunchDialogScript.cs:
```diff
@@ -0,0 +1,63 @@
+﻿using MixedRealityToolkit.InputModule.EventData;
```
Missing license header.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/LaunchDialogScript.cs:
```diff
@@ -0,0 +1,63 @@
+        [SerializeField]
+        private Dialog dialogPrefab;
+
+        public int numButtons;
```
We should probably encapsulate the public fields and use proper PascalCasing for public and camelCasing for private.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,154 @@
+{
+    public class ProgressExamples : MonoBehaviour
+    {
+        public GameObject SceneObject;
```
please encapsulate public fields

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,154 @@
+
+        public void LaunchProgress(GameObject obj)
+        {
+            Debug.Log("Loading with button " + obj.name);
```
Do we really need to throw this debug in production code?

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,154 @@
+            if (HolographicSettings.IsDisplayOpaque)
+            {
+               GameObject buttonCollection = GameObject.Find("ButtonCollection");
+               buttonCollection.transform.localScale = new Vector3(1.3f, 1.3f, 1.0f);
```
Is there a way to customize this scalar value?

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,154 @@
+            // Optimize the content for immersive headset
+            if (HolographicSettings.IsDisplayOpaque)
+            {
+               GameObject buttonCollection = GameObject.Find("ButtonCollection");
```
Where is this button in relation to the GameObject this Component is on?

cre8ivepark replied:
The buttons are under ButtonCollection object in the scene hierarchy.
![2018-02-26 17_17_49-unity 2017 2 1f1 personal 64bit - progressexample unity - mixedrealitytoolkit-](https://user-images.githubusercontent.com/13754172/36705281-26c46442-1b19-11e8-8601-e0eda24d0f00.png)

StephenHodgson replied:
Okay, but where is the `ProgressExamples` component location in relation to this GameObject?

johnppella replied:
i think the scaling of a scene object based on HMD type should be handled by a script that could be added to any UI element that differentially scales. There are two scenes (dialog and progress that use them)That way the object is always known by the instance and doesn't have to be searched for.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,154 @@
+            if (ProgressIndicator.Instance.IsLoading)
+                return;
+
+            switch (obj.name)
```
Not sure if I like that we're switching by the object's name.  What if I want to name my button something else?

johnppella commented:
changed switch(name) to switch on enum for progress indicator example scene
encapsulated, serialized and cased variables
repaired missing header
removed strange spacing in header
removed Debug call 
made UI element scale factor for IHMD vs HoloLens check into a variable.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/LaunchDialogScript.cs:
```diff
@@ -8,13 +11,37 @@ namespace MixedRealityToolkit.Examples.UX
         [SerializeField]
         private Dialog dialogPrefab;
+        private bool isDialogLaunched;
+        private GameObject resultText;
```
if this was public before, do we want to also serialize the field to users can reference the text result object in the inspector?

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/LaunchDialogScript.cs:
```diff
@@ -8,13 +11,37 @@ namespace MixedRealityToolkit.Examples.UX
+            }
+        }
+
+        public int NumButtons
```
Could this be an auto property?

cre8ivepark replied:
Updated to use auto property.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -1,69 +1,306 @@
     public class ProgressExamples : MonoBehaviour
     {
-        public GameObject SceneObject;
+        private GameObject sceneObject;
```
Do we also want to serialize this as well?

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/Prototyping/Shaders/UnlitTransparentTriplanar.shader:
```diff
@@ -1,8 +1,8 @@
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
 ///
-/// Simple vertex shader that blends static lighting and camera lighting.
```
Still a lot of formatting changes in this file.

Could we revert these just to see the real changes please?

cre8ivepark replied:
These files are not related to this update - I have reverted changes on these.

cre8ivepark commented:
@StephenHodgson Applied your feedback. Thanks!

cre8ivepark commented:
- Updated both scenes to assign button collection in the inspector panel.
- Renamed inconsistent button script to avoid confusion. Now they are ProgressExampleLaunchButton.cs and DialogExampleLaunchButton.cs
- Applied same scaling related scripts for the consistency.

StephenHodgson commented on Assets/MixedRealityToolkit/Utilities/Scripts/SceneContentScaler.cs:
```diff
@@ -0,0 +1,57 @@
+
+    }
+
+    private void Update()
```
empty update loop?  We can remove this.

StephenHodgson commented on Assets/MixedRealityToolkit/Utilities/Scripts/SceneContentScaler.cs:
```diff
@@ -0,0 +1,57 @@
+
+public class SceneContentScaler : MonoBehaviour
+{
+    public enum HMDType
```
Do we really need this?  `HolographicSettings.IsDisplayOpaque` should be enough to tell the difference.
Maybe we can have some global value we can set this at, but maybe that's for vNEXT.

cre8ivepark commented:
Resolved merge conflict. Merged Dev_Working_Branch, fixed missing prefabs in the scene.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/DialogExampleLaunchButton.cs:
```diff
@@ -0,0 +1,92 @@
+            }
+        }
+
+        public int NumButtons { get; set; }
```
Does this need to be set in the inspector?
If so we may want to also have a private serialized backing field.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/DialogExampleLaunchButton.cs:
```diff
@@ -0,0 +1,92 @@
+
+        private void OnEnable()
+        {
+            GetComponent<Button>().OnButtonClicked += OnButtonClicked;
```
We should probably require this component or at least check to make sure it's not null and cache it.
Should we also unsubscribe for OnDisable?

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/DialogExampleLaunchButton.cs:
```diff
@@ -0,0 +1,92 @@
+        protected void OnClosed(DialogResult result)
+        {
+            // Get the result text from the Dialog
+            ResultText.GetComponent<TextMesh>().text = result.Result.ToString();
```
We can cache the TextMesh Component too

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/DialogExamples.cs:
```diff
@@ -0,0 +1,32 @@
+
+        private void Start()
+        {
+
```
extra space

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,382 @@
+using System.Collections;
+using UnityEngine;
+
+using UnityEngine.XR.WSA;
```
Doesn't compile for Standalone

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,382 @@
+        private Vector3 scaleIfImmersive = new Vector3(1.3f, 1.3f, 1);
+
+
+        private GameObject sceneObject;
```
Do we want to set this object in the inspector?

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Scripts/ProgressExamples.cs:
```diff
@@ -0,0 +1,382 @@
+                    break;
+            }
+        }
+        protected IEnumerator LoadOverTime(string message)
```
add space after last method

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/Dialog.cs:
```diff
@@ -0,0 +1,214 @@
+        /// <returns></returns>
+        protected IEnumerator RunDialogOverTime()
+        {
+            
```
extra space

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/Dialog.cs:
```diff
@@ -0,0 +1,214 @@
+            // Wait a moment to give scripts a chance to respond
+            yield return null;
+            // Destroy ourselves
+            GameObject.Destroy(gameObject);
```
Do we want to do this, or just disable it so we can reuse it again later?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogButton.cs:
```diff
@@ -0,0 +1,36 @@
+{
+    public class DialogButton : MonoBehaviour
+    {
+        public DialogShell ParentDialog { get; set; }
```
Is this value expected to be set via inspector? If so we need a serialized private backing field.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogButton.cs:
```diff
@@ -0,0 +1,36 @@
+    {
+        public DialogShell ParentDialog { get; set; }
+
+        public Dialog.Dialog.ButtonTypeEnum ButtonTypeEnum;
```
Does this need to be public?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogButton.cs:
```diff
@@ -0,0 +1,36 @@
+
+        private void OnEnable()
+        {
+            GetComponent<Button>().OnButtonClicked += OnButtonClicked;
```
We should probably cache the button component and also unsubscribe on disable.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogButton.cs:
```diff
@@ -0,0 +1,36 @@
+
+        public void SetTitle(string title)
+        {
+            CompoundButtonText c = GetComponent<CompoundButtonText>();
```
Should we also require this component?
Should we also cache it?
Also c is too short a name, `text` would suffice.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogResult.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <summary>
+        /// The button press that closed the dialog
+        /// </summary>
+        public Dialog.ButtonTypeEnum Result = Dialog.ButtonTypeEnum.Close;
```
can this be private? If so please make the change and add the serialized attribute.
If not then add a public property that utilizes the field as a private serialized field.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogShell.cs:
```diff
@@ -0,0 +1,233 @@
+    /// </summary>
+    public class DialogShell : Dialog
+    {
+        public int MaxCharsPerLine = 60;
```
please change this to a private, serialized field.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Dialog/DialogShell.cs:
```diff
@@ -0,0 +1,233 @@
+        protected override void FinalizeLayout()
+        {
+            // Optimize the content for immersive headset
+            if (HolographicSettings.IsDisplayOpaque)
```
we're not wrapping this in a compiler symbol so this would not build for Standanlone as is.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/FadeObject.cs:
```diff
@@ -0,0 +1,127 @@
+﻿using UnityEngine;
```
missing license headers

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/FadeObject.cs:
```diff
@@ -0,0 +1,127 @@
+{
+    public class FadeObject : MonoBehaviour
+    {
+        public float FadeTime = 0.5f;
```
encapsulate public fields

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/FadeObject.cs:
```diff
@@ -0,0 +1,127 @@
+
+namespace MixedRealityToolkit.Examples.InteractiveElements
+{
+    public class FadeObject : MonoBehaviour
```
Don't we already have a class that does something similar to this?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/LoadingAnimation.cs:
```diff
@@ -0,0 +1,217 @@
+﻿using UnityEngine;
```
Missing license headers.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/LoadingAnimation.cs:
```diff
@@ -0,0 +1,217 @@
+﻿using UnityEngine;
+using MixedRealityToolkit.Examples.InteractiveElements;
+
+public class LoadingAnimation : MonoBehaviour
```
missing namspace

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/LoadingAnimation.cs:
```diff
@@ -0,0 +1,217 @@
+
+public class LoadingAnimation : MonoBehaviour
+{
+    public GameObject[] Orbs;
```
please encapsulate public fields

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ObjectDisplayScript.cs:
```diff
@@ -0,0 +1,64 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class ObjectDisplayScript : MonoBehaviour
```
missing namespace

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ObjectDisplayScript.cs:
```diff
@@ -0,0 +1,64 @@
+﻿using System.Collections;
```
missing license headers

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ObjectDisplayScript.cs:
```diff
@@ -0,0 +1,64 @@
+public class ObjectDisplayScript : MonoBehaviour
+{
+    [Header("How fast does object rotate?")]
+    public float RotationIncrement = 200;
```
please encapsulate public fields

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicator.cs:
```diff
@@ -0,0 +1,272 @@
+
+using MixedRealityToolkit.Utilities.Solvers;
+using UnityEngine;
+using UnityEngine.XR.WSA;
```
breaks cross platform support

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicator.cs:
```diff
@@ -0,0 +1,272 @@
+            }
+        }
+
+        public IndicatorStyleEnum DefaultIndicatorStyle = IndicatorStyleEnum.AnimatedOrbs;
```
encapsulation

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicator.cs:
```diff
@@ -0,0 +1,272 @@
+
+            // Make sure we aren't destroyed on load
+            // Just in case the user is loading a scene
+            DontDestroyOnLoad(transform);
```
Singleton class already prevents this from happening. We can safely remove it.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+{
+    public class ProgressIndicatorOrbsRotator : MonoBehaviour
+    {
+        public GameObject[] Orbs;
```
encapsulation of public fields

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+        private float rotationWhenStopped;
+
+        // Use this for initialization
+        void Start()
```
missing private access modifier. Also please remove default comment.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            for (int i = 0; i < Orbs.Length; ++i)
+            {
+                dots[i] = Orbs[i].transform.GetChild(0).gameObject;
+                dots[i].GetComponent<Renderer>().material.color = new Color(1, 1, 1, 1);
```
memory leak when setting color because we don't clean up the created material references.

StephenHodgson replied:
use sharedMaterial

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        // Update is called once per frame
+        void Update()
```
missing private access modifier. Please remove default comment as well.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            rotationWhenStopped = angles[0];
+        }
+
+        void UpdateTime()
```
missing access modifier.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            }
+        }
+
+        void ControlDotStarts()
```
missing access modifier.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            }
+        }
+
+        void IncrementOrbs()
```
missing access modifier.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            }
+        }
+
+        void IncrementOrb(int index)
```
missing access modifier.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            HandleFade(index);
+        }
+
+        void HandleFade(int index)
```
missing access modifier.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+
+        void HandleFade(int index)
+        {
+            Color adjustedColor = dots[index].GetComponent<Renderer>().material.color;
```
memory leak

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            {
+                adjustedColor.a += (1.0f * timeSlice);
+                adjustedColor.a = Mathf.Min(1.0f, adjustedColor.a);
+                dots[index].GetComponent<Renderer>().material.color = adjustedColor;
```
memory leak

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            {
+                adjustedColor.a -= (1.0f * timeSlice);
+                adjustedColor.a = Mathf.Max(0.0f, adjustedColor.a);
+                dots[index].GetComponent<Renderer>().material.color = adjustedColor;
```
memory leak

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            }
+        }
+
+        void HandleTestStop()
```
missing access modifier.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+            }
+        }
+
+        void HandleStopping()
```
missing access modifier.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Progress/ProgressIndicatorOrbsRotator.cs:
```diff
@@ -0,0 +1,156 @@
+
+        void HandleStopping()
+        {
+            if (stopRequested == true && dots[Orbs.Length - 1].GetComponent<Renderer>().material.color.a <= 0.01f)
```
memory leak

johnppella commented:
many convention conformance edits.
-encapsulation
-adding missing access modifiers

-removing or enclosing within #ifdefs referencing  UnityEngine.XR.WSA
-addressed Renderer.Material access memory leak by using sharedMaterial

cre8ivepark commented:
@StephenHodgson README conflict Resolved.
